bookdown/03-optimization.Rmd

```diff
@@ -231,6 +231,231 @@ Note that we do not expect any differences here compared to the non-tuned approa
+Hyperband is a budget-oriented procedure, weeding out suboptimally performing configurations early on during a partially sequential training process, increasing tuning efficiency as a consequence.
+For this, a combination of incremental resource allocation and early stopping is used, increasing computational resources of more promising configurations plus terminating lesser ones the further training stages are reached.
+To give an introductional analogy, imagine two horse trainers are given eight untrained horses.
+Both trainers want to end up with the quickest horse to win the incoming race, but they are only given a fixed amount of budget (32 units) to buy food.
+But, because the food is not enough for all the horses, it is critical to give the right ones enough food to improve.
+So, the trainers need to develop a strategy to split up the food between the horses.
+The first trainer is very optimistic and wants to explore the full capabilities of a horse, because he does not want to a judgement before they are not fully trained.
+So, he divides his budget by the maximum amount he can give to a horse (lets say eigth, so $32 / 8 = 4$) and randomly picks four horses -- his budget simply is not enough to fully train more.
+Those four horses are then trained to their full capabilities, while the rest is set free.
+The other trainer is smarter but also very pessimistic and exploitative.
+He thinks, if a horse is not performing well at the beginning, it will also not improve after further training.
+Based on this assumption he develops a strategy that only gives very little food to the horses at first.
+After the initial food is consumed, he checks their performance and kicks the slowest half out of his training regime.
+Then, he increases the available food for the remaining, further trains them until the food is consumed again, only to kick out the worst half once more.
+He repeats this until the one remaining horse gets the rest of the food.
+But, on the flip side, he was able to start training with all eight horses.
+On race day, all the horses are put on the starting line.
+But which trainer will have the winning horse?
+The one, who tried to train a maximum amount of horses to their fullest?
+Or the other one, who is exploiting the training progress of his horses?
+Hyperband works very similar in some ways, but also different in others.
+It is not embodied by one of the trainers in our analogy, but more by the person, who would pay them.
+Hyperband consists of several brackets, each bracket corresponding to a trainer, and we do not care about horses but about hyperparameter configurations of a machine learning algorithm.
+The budget is not in terms of food, but in terms of a hyperparameter of the learner that scales in some way with the computational effort.
+An example is the epoch amount of neural networks, or the tree amount in (tree-based) boosting.
```

Comment:
[...] the number of epochs we train a neural network, or the number of iterations in [...]